extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -0,0 +1,705 @@
+      return Collections.emptyList();
+    }
+
+  }
+
+  @Override
+  public String getLatestSequenceNumber(StreamPartition<String> partition)
+  {
+    checkIfClosed();
+    return getSequenceNumberInternal(partition, ShardIteratorType.LATEST);
+  }
+
+  @Override
+  public String getEarliestSequenceNumber(StreamPartition<String> partition)
+  {
+    checkIfClosed();
+    return getSequenceNumberInternal(partition, ShardIteratorType.TRIM_HORIZON);
+  }
+
+  @Nullable
+  @Override
+  public String getPosition(StreamPartition<String> partition)
+  {
+    throw new UnsupportedOperationException("getPosition is not supported in Kinesiss");
+  }
+
+  @Override
+  public Set<String> getPartitionIds(String stream)
+  {
+    checkIfClosed();
+    return kinesis.describeStream(stream)
+                  .getStreamDescription()
+                  .getShards()
+                  .stream()
+                  .map(Shard::getShardId).collect(Collectors.toSet());
+  }
+
+  @Override
+  public void close()
+  {
+    if (this.closed) {
+      return;
+    }
+
+    assign(ImmutableSet.of());
+
+    scheduledExec.shutdown();
+
+    try {
+      if (!scheduledExec.awaitTermination(EXCEPTION_RETRY_DELAY_MS, TimeUnit.MILLISECONDS)) {
+        scheduledExec.shutdownNow();
+      }
+    }
+    catch (InterruptedException e) {
+      log.info(e, "InterruptedException while shutting down");
+    }
+
+    this.closed = true;
+  }
+
+  private void seekInternal(StreamPartition<String> partition, String sequenceNumber, ShardIteratorType iteratorEnum)
+  {
+    PartitionResource resource = partitionResources.get(partition);
+    if (resource == null) {
+      throw new ISE("Partition [%s] has not been assigned", partition);
+    }
+
+    log.debug(
+        "Seeking partition [%s] to [%s]",
+        partition.getPartitionId(),
+        sequenceNumber != null ? sequenceNumber : iteratorEnum.toString()
+    );
+
+    resource.shardIterator = kinesis.getShardIterator(
+        partition.getStream(),
+        partition.getPartitionId(),
+        iteratorEnum.toString(),
+        sequenceNumber
+    ).getShardIterator();
+
+    checkPartitionsStarted = true;
+  }
+
+  private void filterBufferAndResetFetchRunnable(Set<StreamPartition<String>> partitions)
+  {
+    scheduledExec.shutdown();
+
+    try {
+      if (!scheduledExec.awaitTermination(EXCEPTION_RETRY_DELAY_MS, TimeUnit.MILLISECONDS)) {
+        scheduledExec.shutdownNow();
+      }
+    }
+    catch (InterruptedException e) {
+      log.info(e, "InterruptedException while shutting down");
+    }
+
+    scheduledExec = Executors.newScheduledThreadPool(
+        fetchThreads,
+        Execs.makeThreadFactory("KinesisRecordSupplier-Worker-%d")
+    );
+
+    // filter records in buffer and only retain ones whose partition was not seeked
+    BlockingQueue<OrderedPartitionableRecord<String, String>> newQ = new LinkedBlockingQueue<>(recordBufferSize);
+    records
+        .parallelStream()
```

Comment:
Does this guarantee the item order in `records` is always same in `newQ`?

Reply:
yes strict ordering is guaranteed under `parallelStream` https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html#Ordering

Reply:
Thanks. But, I'm still not sure we need parallel stream here. Since # of CPU cores is limited, we should be careful when we use them. This `recordSupplier` is used in indexTask which is also responsible for query processing and I think it's more important to make the query processing faster than fast ingestion. What do you think?

Reply:
makes sense, I just thought that we expect there to be many records in `records` to filter so it might make sense to improve the filtering operation, but I guess it's not really needed hehe